In [1]:
# import libraries
import pandas as pd
import json

In [2]:
# load data
df_info = pd.read_csv('Data/superhero_info - superhero_info.csv')
df_powers = pd.read_csv('Data/superhero_powers - superhero_powers.csv')

In [3]:
df_info.head()

Hero|Publisher Gender               Race Alignment  \
0          A-Bomb|Marvel Comics   Male              Human      good   
1  Abe Sapien|Dark Horse Comics   Male      Icthyo Sapien      good   
2            Abin Sur|DC Comics   Male            Ungaran      good   
3     Abomination|Marvel Comics   Male  Human / Radiation       bad   
4   Absorbing Man|Marvel Comics   Male              Human       bad   

  Hair color Eye color Skin color  \
0    No Hair    yellow    Unknown   
1    No Hair      blue       blue   
2    No Hair      blue        red   
3    No Hair     green    Unknown   
4    No Hair      blue    Unknown   

                                   Measurements  
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}  
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}

In [4]:
df_powers.head()

hero_names                                             Powers
0      3-D Man         Agility,Super Strength,Stamina,Super Speed
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...
3     Abin Sur                                 Lantern Power Ring
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...

## Cleaning df_info

In [5]:
# split Hero and Publisher
df_info['Hero|Publisher'].str.split('|', expand=True)

0                  1
0             A-Bomb      Marvel Comics
1         Abe Sapien  Dark Horse Comics
2           Abin Sur          DC Comics
3        Abomination      Marvel Comics
4      Absorbing Man      Marvel Comics
..               ...                ...
458     Yellowjacket      Marvel Comics
459  Yellowjacket II      Marvel Comics
460             Yoda       George Lucas
461          Zatanna          DC Comics
462             Zoom          DC Comics

[463 rows x 2 columns]

In [6]:
df_info[['Hero','Publisher']] = df_info['Hero|Publisher'].str.split('|', expand=True)
df_info.head()

Hero|Publisher Gender               Race Alignment  \
0          A-Bomb|Marvel Comics   Male              Human      good   
1  Abe Sapien|Dark Horse Comics   Male      Icthyo Sapien      good   
2            Abin Sur|DC Comics   Male            Ungaran      good   
3     Abomination|Marvel Comics   Male  Human / Radiation       bad   
4   Absorbing Man|Marvel Comics   Male              Human       bad   

  Hair color Eye color Skin color  \
0    No Hair    yellow    Unknown   
1    No Hair      blue       blue   
2    No Hair      blue        red   
3    No Hair     green    Unknown   
4    No Hair      blue    Unknown   

                                   Measurements           Hero  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher  
0      Marvel Comics  
1  Dark Horse Comics  
2          DC Comics  
3      Marvel Comics  
4      Marvel Comics

In [7]:
df_info = df_info.drop(columns=['Hero|Publisher'])

In [8]:
# convert measurements to dictionary and extract measures height and weight in separate columns
measure = df_info.loc[0,'Measurements']
print(type(measure))
measure

<class 'str'>


"{'Height': '203.0 cm', 'Weight': '441.0 kg'}"

In [9]:
# replace ' woith '
df_info['Measurements'] = df_info['Measurements'].str.replace("'",'"')
df_info['Measurements'].head()

0    {"Height": "203.0 cm", "Weight": "441.0 kg"}
1     {"Height": "191.0 cm", "Weight": "65.0 kg"}
2     {"Height": "185.0 cm", "Weight": "90.0 kg"}
3    {"Height": "203.0 cm", "Weight": "441.0 kg"}
4    {"Height": "193.0 cm", "Weight": "122.0 kg"}
Name: Measurements, dtype: object

In [10]:
# convert to dictionary
df_info['Measurements'] = df_info['Measurements'].apply(json.loads)

In [11]:
df_info['Measurements'].head()

0    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
1     {'Height': '191.0 cm', 'Weight': '65.0 kg'}
2     {'Height': '185.0 cm', 'Weight': '90.0 kg'}
3    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
4    {'Height': '193.0 cm', 'Weight': '122.0 kg'}
Name: Measurements, dtype: object

In [12]:
measure = df_info.loc[0,'Measurements']
print(type(measure))
measure

<class 'dict'>


{'Height': '203.0 cm', 'Weight': '441.0 kg'}

In [13]:
# convert dictionary to dataframe with keys as column names
height_weight = df_info['Measurements'].apply(pd.Series)
height_weight

Height    Weight
0    203.0 cm  441.0 kg
1    191.0 cm   65.0 kg
2    185.0 cm   90.0 kg
3    203.0 cm  441.0 kg
4    193.0 cm  122.0 kg
..        ...       ...
458  183.0 cm   83.0 kg
459  165.0 cm   52.0 kg
460   66.0 cm   17.0 kg
461  170.0 cm   57.0 kg
462  185.0 cm   81.0 kg

[463 rows x 2 columns]

In [14]:
# concatenate dataframes
df_info = pd.concat((df_info,height_weight), axis=1)
df_info.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

                                   Measurements           Hero  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher    Height    Weight  
0      Marvel Comics  203.0 cm  441.0 kg  
1  Dark Horse Comics  191.0 cm   65.0 kg  
2          DC Comics  185.0 cm   90.0 kg  
3      Marvel Comics  203.0 cm  441.0 kg  
4      Marvel Comics  193.0 cm  122.0 kg

In [15]:
# remove units from Height and Weight columns
df_info['Height'] = df_info['Height'].str.replace(' cm','')
df_info['Weight'] = df_info['Weight'].str.replace(' kg','')
df_info.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

                                   Measurements           Hero  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher Height Weight  
0      Marvel Comics  203.0  441.0  
1  Dark Horse Comics  191.0   65.0  
2          DC Comics  185.0   90.0  
3      Marvel Comics  203.0  441.0  
4      Marvel Comics  193.0  122.0

In [16]:
df_info['Height'].dtype

dtype('O')

In [17]:
df_info['Height'] = df_info['Height'].astype(float)
df_info['Height'].dtype

dtype('float64')

In [18]:
df_info['Weight'].dtype

dtype('O')

In [19]:
df_info['Weight'] = df_info['Weight'].astype(float)
df_info['Weight'].dtype

dtype('float64')

In [20]:
df_info.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

                                   Measurements           Hero  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher  Height  Weight  
0      Marvel Comics   203.0   441.0  
1  Dark Horse Comics   191.0    65.0  
2          DC Comics   185.0    90.0  
3      Marvel Comics   203.0   441.0  
4      Marvel Comics   193.0   122.0

In [21]:
df_info = df_info.drop(columns='Measurements')

In [22]:
df_info['Hero'].nunique()

455

## Clean df_powers

In [23]:
df_powers.head()

hero_names                                             Powers
0      3-D Man         Agility,Super Strength,Stamina,Super Speed
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...
3     Abin Sur                                 Lantern Power Ring
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...

In [24]:
power = df_powers.loc[0,'Powers']
print(type(power))
power

<class 'str'>


'Agility,Super Strength,Stamina,Super Speed'

In [25]:
# convert string to list format
df_powers['Powers_list'] = '["'+df_powers['Powers'].str.replace(',','","')+'"]'
df_powers.head()

hero_names                                             Powers  \
0      3-D Man         Agility,Super Strength,Stamina,Super Speed   
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
3     Abin Sur                                 Lantern Power Ring   
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...   

                                         Powers_list  
0  ["Agility","Super Strength","Stamina","Super S...  
1  ["Accelerated Healing","Durability","Longevity...  
2  ["Agility","Accelerated Healing","Cold Resista...  
3                             ["Lantern Power Ring"]  
4  ["Accelerated Healing","Intelligence","Super S...

In [26]:
# convert to list
df_powers['Powers_list'] = df_powers['Powers_list'].apply(json.loads)

df_powers['Powers_list'].head()

0      [Agility, Super Strength, Stamina, Super Speed]
1    [Accelerated Healing, Durability, Longevity, S...
2    [Agility, Accelerated Healing, Cold Resistance...
3                                 [Lantern Power Ring]
4    [Accelerated Healing, Intelligence, Super Stre...
Name: Powers_list, dtype: object

In [27]:
power = df_powers.loc[0,'Powers_list']
print(type(power))
power

<class 'list'>


['Agility', 'Super Strength', 'Stamina', 'Super Speed']

In [28]:
# explode list
exploded = df_powers.explode('Powers_list')
exploded.head()

hero_names                                             Powers  \
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
1     A-Bomb  Accelerated Healing,Durability,Longevity,Super...   

           Powers_list  
0              Agility  
0       Super Strength  
0              Stamina  
0          Super Speed  
1  Accelerated Healing

In [29]:
cols_to_make = exploded['Powers_list'].unique()
cols_to_make

array(['Agility', 'Super Strength', 'Stamina', 'Super Speed',
       'Accelerated Healing', 'Durability', 'Longevity', 'Camouflage',
       'Self-Sustenance', 'Cold Resistance', 'Underwater breathing',
       'Marksmanship', 'Weapons Master', 'Intelligence', 'Telepathy',
       'Immortality', 'Reflexes', 'Enhanced Sight', 'Sub-Mariner',
       'Lantern Power Ring', 'Invulnerability', 'Animation',
       'Super Breath', 'Dimensional Awareness', 'Flight', 'Size Changing',
       'Teleportation', 'Magic', 'Dimensional Travel',
       'Molecular Manipulation', 'Energy Manipulation', 'Power Cosmic',
       'Energy Absorption', 'Elemental Transmogrification',
       'Fire Resistance', 'Natural Armor', 'Heat Resistance',
       'Matter Absorption', 'Regeneration', 'Stealth', 'Power Suit',
       'Energy Blasts', 'Energy Beams', 'Heat Generation', 'Danger Sense',
       'Phasing', 'Force Fields', 'Hypnokinesis', 'Invisibility',
       'Enhanced Senses', 'Jump', 'Shapeshifting', 'Elasticity',
 

In [30]:
# create new column for each power
for col in cols_to_make:
    df_col = df_powers.copy()
    df_col[col] = df_col['Powers'].str.contains(col)
    df_powers = pd.concat((df_powers,df_col[col]), axis=1)
df_powers.head()

hero_names                                             Powers  \
0      3-D Man         Agility,Super Strength,Stamina,Super Speed   
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
3     Abin Sur                                 Lantern Power Ring   
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...   

                                         Powers_list  Agility  Super Strength  \
0    [Agility, Super Strength, Stamina, Super Speed]     True            True   
1  [Accelerated Healing, Durability, Longevity, S...    False            True   
2  [Agility, Accelerated Healing, Cold Resistance...     True            True   
3                               [Lantern Power Ring]    False           False   
4  [Accelerated Healing, Intelligence, Super Stre...    False            True   

   Stamina  Super Speed  Accelerated Healing  Durability  Longevity  ...  \
0     True         True                False       False      False  ...   
1     True        False                 True        True       True  ...   
2     True        False                 True        True       True  ...   
3    False        False                False       False      False  ...   
4     True         True                 True       False      False  ...   

   Weather Control  Omnipresent  Omniscient  Hair Manipulation  Nova Force  \
0            False        False       False              False       False   
1            False        False       False              False       False   
2            False        False       False              False       False   
3            False        False       False              False       False   
4            False        False       False              False       False   

   Odin Force  Phoenix Force  Intuitive aptitude  Melting  Changing Armor  
0       False          False               False    False           False  
1       False          False               False    False           False  
2       False          False               False    False           False  
3       False          False               False    False           False  
4       False          False               False    False           False  

[5 rows x 170 columns]

In [31]:
df_powers['hero_names'].nunique()

667

In [32]:
df_powers = df_powers.drop(columns=['Powers','Powers_list'])
df_powers.head()

hero_names  Agility  Super Strength  Stamina  Super Speed  \
0      3-D Man     True            True     True         True   
1       A-Bomb    False            True     True        False   
2   Abe Sapien     True            True     True        False   
3     Abin Sur    False           False    False        False   
4  Abomination    False            True     True         True   

   Accelerated Healing  Durability  Longevity  Camouflage  Self-Sustenance  \
0                False       False      False       False            False   
1                 True        True       True        True             True   
2                 True        True       True       False            False   
3                False       False      False       False            False   
4                 True       False      False       False            False   

   ...  Weather Control  Omnipresent  Omniscient  Hair Manipulation  \
0  ...            False        False       False              False   
1  ...            False        False       False              False   
2  ...            False        False       False              False   
3  ...            False        False       False              False   
4  ...            False        False       False              False   

   Nova Force  Odin Force  Phoenix Force  Intuitive aptitude  Melting  \
0       False       False          False               False    False   
1       False       False          False               False    False   
2       False       False          False               False    False   
3       False       False          False               False    False   
4       False       False          False               False    False   

   Changing Armor  
0           False  
1           False  
2           False  
3           False  
4           False  

[5 rows x 168 columns]

## Merge df_info and df_powers

In [33]:
df_final = pd.merge(df_info, df_powers, left_on='Hero', right_on='hero_names')

In [34]:
df_final.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

            Hero          Publisher  Height  Weight  ... Weather Control  \
0         A-Bomb      Marvel Comics   203.0   441.0  ...           False   
1     Abe Sapien  Dark Horse Comics   191.0    65.0  ...           False   
2       Abin Sur          DC Comics   185.0    90.0  ...           False   
3    Abomination      Marvel Comics   203.0   441.0  ...           False   
4  Absorbing Man      Marvel Comics   193.0   122.0  ...           False   

   Omnipresent  Omniscient  Hair Manipulation  Nova Force  Odin Force  \
0        False       False              False       False       False   
1        False       False              False       False       False   
2        False       False              False       False       False   
3        False       False              False       False       False   
4        False       False              False       False       False   

   Phoenix Force  Intuitive aptitude  Melting  Changing Armor  
0          False               False    False           False  
1          False               False    False           False  
2          False               False    False           False  
3          False               False    False           False  
4          False               False    False           False  

[5 rows x 178 columns]

In [35]:
df_final.columns

Index(['Gender', 'Race', 'Alignment', 'Hair color', 'Eye color', 'Skin color',
       'Hero', 'Publisher', 'Height', 'Weight',
       ...
       'Weather Control', 'Omnipresent', 'Omniscient', 'Hair Manipulation',
       'Nova Force', 'Odin Force', 'Phoenix Force', 'Intuitive aptitude',
       'Melting', 'Changing Armor'],
      dtype='object', length=178)

In [36]:
df_final = df_final.drop(columns='hero_names')

In [37]:
df_final['Hero'].nunique()

455

Compare the average weight of super powers who have Super Speed to those who do not.

In [38]:
df_final.groupby('Super Speed')['Weight'].mean()

Super Speed
False    101.773585
True     129.404040
Name: Weight, dtype: float64

The average weight of Heros who have Super Speed is 129.4 kg's while those that do not have Super Speed is 101.8 kg's

What is the average height of heroes for each publisher?

In [39]:
df_final.groupby('Publisher')['Height'].mean()

Publisher
DC Comics            181.923913
Dark Horse Comics    176.909091
George Lucas         159.600000
Image Comics         211.000000
Marvel Comics        191.546128
Shueisha             171.500000
Star Trek            181.500000
Team Epic TV         180.750000
Unknown              178.000000
Name: Height, dtype: float64